source/logrepl/internal/subscription.go

```diff
@@ -27,125 +27,119 @@ import (
 }
 
 type Handler func(context.Context, pglogrepl.Message, pglogrepl.LSN) error
 
-func NewSubscription(
-	config pgconn.Config,
+func CreateSubscription(
+	ctx context.Context,
+	conn *pgconn.PgConn,
 	slotName,
 	publication string,
 	tables []string,
 	startLSN pglogrepl.LSN,
 	h Handler,
-) *Subscription {
+) (*Subscription, error) {
+	result, err := pglogrepl.CreateReplicationSlot(
+		ctx,
+		conn,
+		slotName,
+		pgOutputPlugin,
+		pglogrepl.CreateReplicationSlotOptions{
+			SnapshotAction: "EXPORT_SNAPSHOT",
+			Mode:           pglogrepl.LogicalReplication,
+		},
+	)
+	if err != nil {
+		// If creating the replication slot fails with code 42710, this means
+		// the replication slot already exists.
+		if !IsPgDuplicateErr(err) {
+			return nil, err
+		}
+
+		sdk.Logger(ctx).Warn().
+			Msgf("replication slot %q already exists", slotName)
+	}
+
 	return &Subscription{
-		ConnConfig:    config,
 		SlotName:      slotName,
 		Publication:   publication,
 		Tables:        tables,
 		StartLSN:      startLSN,
 		Handler:       h,
 		StatusTimeout: 10 * time.Second,
+		TXSnapshotID:  result.SnapshotName,
+
+		conn: conn,
 
 		ready: make(chan struct{}),
 		done:  make(chan struct{}),
-		// cleanup does nothing by default
-		cleanup: func(context.Context) error { return nil },
-	}
+	}, nil
 }
 
 // Start replication and block until error or ctx is canceled.
-func (s *Subscription) Start(ctx context.Context) (err error) {
-	defer func() {
-		// use fresh context for cleanup
-		cleanupErr := s.cleanup(context.Background())
-		if err == nil {
-			// return close connection error
-			err = cleanupErr
-		} else if cleanupErr != nil {
-			// an error is already returned, let's log this one instead
-			sdk.Logger(ctx).Err(cleanupErr).Msg("failed to cleanup subscription")
-		}
-		s.doneErr = err // store error so it can be retrieved later
-
-		select {
-		case <-s.ready:
-			// ready is already closed
-		default:
-			close(s.ready)
-		}
-		close(s.done)
-	}()
-
-	conn, err := s.connect(ctx)
-	if err != nil {
-		return err
-	}
-	err = s.createPublication(ctx, conn)
-	if err != nil {
-		return err
-	}
-	err = s.createReplicationSlot(ctx, conn)
-	if err != nil {
-		return err
-	}
-	err = s.startReplication(ctx, conn)
-	if err != nil {
+func (s *Subscription) Start(ctx context.Context) error {
+	if err := s.startReplication(ctx); err != nil {
 		return err
 	}
 
 	lctx, cancel := context.WithCancel(ctx)
-	defer cancel()
 	s.stop = cancel
 	s.walWritten = s.StartLSN
 	s.walFlushed = s.StartLSN
 
-	return s.listen(lctx, conn)
+	go func() {
+		if err := s.listen(lctx); err != nil {
+			sdk.Logger(ctx).Error().
+				Err(err).
+				Msg("replication exited with an error")
+		}
+	}()
+
+	<-s.ready
+
+	return nil
 }
 
 // listen runs until context is cancelled or an error is encountered.
-func (s *Subscription) listen(ctx context.Context, conn *pgconn.PgConn) error {
+func (s *Subscription) listen(ctx context.Context) error {
+	defer s.doneReplication()
```

Comment:
It might help if you give `ctx` to `s.doneReplication()`, since that method uses `context.Background()`.

Reply:
this was on purpose, so replication clean up cannot be cancelled, otherwise it will be half done.

Reply:
see f28285404c22064f4ac63af063fcc785358965c3, timeout added